File: packages/ucm-core/src/markers/bindingSetHash.ts
import { canonicalJsonSha256 } from "./canonicalJson.js";
import { BINDING_SET_HASH_ID } from "./constants.js";

// A single binding as seen by the scanner. Extra diagnostic fields (line
// numbers, byte offsets, markers, etc.) are permitted on the input but are
// deliberately excluded from the hash material per spec section 4.5.
export interface BindingSetInputMember {
  binding_slug: string;
  row_id: string;
  file_path: string;
  extent_kind: string;
  recognizer_id: string;
  span_canon_id: string;
  span_sha256: string;
  [key: string]: unknown;
}

// The exact, hashed subset of a binding (whitelist; spec section 4.5).
export interface BindingSetMaterialMember {
  binding_slug: string;
  row_id: string;
  file_path: string;
  extent_kind: string;
  recognizer_id: string;
  span_canon_id: string;
  span_sha256: string;
}

export interface BindingSetMaterial {
  schema: typeof BINDING_SET_HASH_ID;
  row_id: string;
  bindings: BindingSetMaterialMember[];
}

// Build the canonical binding-set material for a row.
//
// Included: binding_slug, row_id, file_path, extent_kind, recognizer_id,
// span_canon_id, span_sha256 (spec 4.5 rules 2-6).
// Excluded: line numbers, timestamps, proof status, marker text (rules 7-10).
// Bindings are sorted by binding_slug before hashing (rule 1).
export function buildBindingSetMaterial(
  rowId: string,
  bindings: ReadonlyArray<BindingSetInputMember>
): BindingSetMaterial {
  const sorted = [...bindings].sort((left, right) =>
    left.binding_slug < right.binding_slug
      ? -1
      : left.binding_slug > right.binding_slug
        ? 1
        : 0
  );
  return {
    schema: BINDING_SET_HASH_ID,
    row_id: rowId,
    bindings: sorted.map((binding) => ({
      binding_slug: binding.binding_slug,
      row_id: binding.row_id,
      file_path: binding.file_path,
      extent_kind: binding.extent_kind,
      recognizer_id: binding.recognizer_id,
      span_canon_id: binding.span_canon_id,
      span_sha256: binding.span_sha256
    }))
  };
}

// binding_set_hash = sha256(canonical_json(binding_set_material)).
export function computeBindingSetHash(
  rowId: string,
  bindings: ReadonlyArray<BindingSetInputMember>
): string {
  return canonicalJsonSha256(buildBindingSetMaterial(rowId, bindings));
}

File: UseCasesCore/Sources/UseCasesCore/Markers/BindingSetHash.swift
/// One binding as the binding-set hash sees it: exactly the seven hashed
/// fields (spec 4.5 rules 2–6). Line numbers, markers and proof status are
/// never part of it (rules 7–10).
public struct BindingSetMember: Equatable, Sendable {
  public let bindingSlug: String
  public let rowIdentifier: String
  public let filePath: String
  public let extentKind: String
  public let recognizerIdentifier: String
  public let spanCanonicalizerIdentifier: String
  public let spanSHA256: String

  public init(
    bindingSlug: String,
    rowIdentifier: String,
    filePath: String,
    extentKind: String,
    recognizerIdentifier: String,
    spanCanonicalizerIdentifier: String,
    spanSHA256: String,
  ) {
    self.bindingSlug = bindingSlug
    self.rowIdentifier = rowIdentifier
    self.filePath = filePath
    self.extentKind = extentKind
    self.recognizerIdentifier = recognizerIdentifier
    self.spanCanonicalizerIdentifier = spanCanonicalizerIdentifier
    self.spanSHA256 = spanSHA256
  }

  /// The hashed fields of a binding item as written in a proof event; any
  /// other member of the item is ignored. Nil when a hashed field is not a
  /// string.
  public init?(json: JSONValue) {
    guard let bindingSlug = json["binding_slug"]?.stringValue,
          let rowIdentifier = json["row_id"]?.stringValue,
          let filePath = json["file_path"]?.stringValue,
          let extentKind = json["extent_kind"]?.stringValue,
          let recognizerIdentifier = json["recognizer_id"]?.stringValue,
          let spanCanonicalizerIdentifier = json["span_canon_id"]?.stringValue,
          let spanSHA256 = json["span_sha256"]?.stringValue
    else {
      return nil
    }
    self.init(
      bindingSlug: bindingSlug,
      rowIdentifier: rowIdentifier,
      filePath: filePath,
      extentKind: extentKind,
      recognizerIdentifier: recognizerIdentifier,
      spanCanonicalizerIdentifier: spanCanonicalizerIdentifier,
      spanSHA256: spanSHA256,
    )
  }

  var jsonValue: JSONValue {
    .object(JSONObject([
      ("binding_slug", .string(bindingSlug)),
      ("row_id", .string(rowIdentifier)),
      ("file_path", .string(filePath)),
      ("extent_kind", .string(extentKind)),
      ("recognizer_id", .string(recognizerIdentifier)),
      ("span_canon_id", .string(spanCanonicalizerIdentifier)),
      ("span_sha256", .string(spanSHA256)),
    ]))
  }
}

/// `binding_set_hash = sha256(canonical_json(binding_set_material))`
/// (bindingSetHash.ts, spec 4.5).
public enum BindingSetHash {
  /// `{ schema, row_id, bindings }`, with the bindings sorted by slug in
  /// UTF-16 code-unit order (rule 1). Swift's own `String` ordering is
  /// canonical equivalence, under which `"é"` and `"e\u{301}"` are EQUAL; the
  /// TypeScript's `<` is not. Equal slugs keep their input order, as
  /// JavaScript's stable sort keeps them.
  public static func material(
    rowIdentifier: String,
    bindings: [BindingSetMember],
  ) -> JSONValue {
    let sorted = bindings.enumerated().sorted { left, right in
      if JavaScriptString.precedes(left.element.bindingSlug, right.element.bindingSlug) {
        return true
      }
      if JavaScriptString.precedes(right.element.bindingSlug, left.element.bindingSlug) {
        return false
      }
      return left.offset < right.offset
    }
    return .object(JSONObject([
      ("schema", .string(MarkerConstants.bindingSetHashIdentifier)),
      ("row_id", .string(rowIdentifier)),
      ("bindings", .array(sorted.map(\.element.jsonValue))),
    ]))
  }

  public static func compute(
    rowIdentifier: String,
    bindings: [BindingSetMember],
  ) throws(CodeUnitCanonicalJSONError) -> String {
    try CodeUnitCanonicalJSON.sha256(material(rowIdentifier: rowIdentifier, bindings: bindings))
  }
}
